File: benchmarks/scala/fib.scala
object fib {

  def normalFib(n: Int): Int =
    if (n <= 1)
      n
    else
      normalFib(n - 1) + normalFib(n - 2)

  def tailFib(n: Int): Int = {
    def fibHelper(number: Int, x: Int, y: Int): Int = {
      if (number == 0)
        x
      else
        fibHelper(number - 1, y, x + y)
    }
    fibHelper(n, 0, 1)
  }

  def main(args: Array[String]): Unit = {
    val number = args(0).toInt
    if (args.length > 1)
      if (args(1) == "tail")
        println(tailFib(number))
      else if (args(1) == "norm")
        println(normalFib(number))
      else
        println("Invalid arguments: wrong fib type")
    else
      println("Invalid arguments: no fib type")
  }
}

File: benchmarks/haskell/fib.hs
import System.Environment (getArgs)

main :: IO ()
main = do
  args <- getArgs
  case args of
    number : "norm" : rest -> print $ normFib (read number)
    number : "tail" : rest -> print $ tailFib (read number)
    number : "memo" : rest -> print $ memoFib (read number)
    [] -> putStrLn "You need to provide a number"
    _ -> putStrLn "Invalid arguments"

tailFib :: (Eq t1, Num t1, Num t2) => t1 -> t2
tailFib n = fibHelper n 0 1
  where
    fibHelper 0 x _ = x
    fibHelper n x y = fibHelper (n -1) y (x + y)

normFib :: (Eq a, Num a, Num p) => a -> p
normFib n = fibHelper n
  where
    fibHelper 0 = 0
    fibHelper 1 = 1
    fibHelper n = fibHelper (n - 1) + fibHelper (n - 2)

memoFib :: Int -> Integer
memoFib = (map fib [0 ..] !!)
  where
    fib 0 = 0
    fib 1 = 1
    fib n = memoFib (n -2) + memoFib (n -1)
